ClassCastException due to HashSet being used in a non thread-safe way

**Jetty version(s)**
12.0.24

**Jetty Environment**
ee10

**Java version/vendor** `(use: java -version)`
openjdk version "21.0.7" 2025-04-15 LTS
OpenJDK Runtime Environment Temurin-21.0.7+6 (build 21.0.7+6-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.7+6 (build 21.0.7+6-LTS, mixed mode, sharing)

**OS type/version**
Windows 11

**Description**
Sometimes we get the following error upon startup:

```
Error scanning entry java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')

Ursprüngliche Fehlerbeschreibung:
(java.lang.ClassCastException)
class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
	java.lang.RuntimeException: Error scanning entry java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
		at org.eclipse.jetty.ee10.annotations.AnnotationParser.parseDir(AnnotationParser.java:613)
		at org.eclipse.jetty.ee10.annotations.AnnotationParser.parse(AnnotationParser.java:561)
		at org.eclipse.jetty.ee10.annotations.AnnotationConfiguration$ParserTask.call(AnnotationConfiguration.java:163)
		at org.eclipse.jetty.util.ExceptionUtil.call(ExceptionUtil.java:339)
		at org.eclipse.jetty.util.ExceptionUtil$MultiException.callAndCatch(ExceptionUtil.java:277)
		at org.eclipse.jetty.ee10.annotations.AnnotationConfiguration.lambda$scanForAnnotations$0(AnnotationConfiguration.java:547)
		at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
		at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
		at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
		at java.base/java.lang.Thread.run(Thread.java:1583)
	Caused by: java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')
		at java.base/java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1994)
		at java.base/java.util.HashMap$TreeNode.treeify(HashMap.java:2110)
		at java.base/java.util.HashMap.treeifyBin(HashMap.java:778)
		at java.base/java.util.HashMap.putVal(HashMap.java:650)
		at java.base/java.util.HashMap.put(HashMap.java:618)
		at java.base/java.util.HashSet.add(HashSet.java:229)
		at java.base/java.util.AbstractCollection.addAll(AbstractCollection.java:338)
		at org.eclipse.jetty.ee10.annotations.AnnotationConfiguration$DiscoveredServletContainerInitializerHolder.addStartupClasses(AnnotationConfiguration.java:336)
		at org.eclipse.jetty.ee10.annotations.ContainerInitializerAnnotationHandler.handle(ContainerInitializerAnnotationHandler.java:49)
		at org.eclipse.jetty.ee10.annotations.AnnotationParser$MyClassVisitor.visitAnnotation(AnnotationParser.java:491)
		at org.objectweb.asm.ClassReader.accept(ClassReader.java:610)
		at org.objectweb.asm.ClassReader.accept(ClassReader.java:425)
		at org.eclipse.jetty.ee10.annotations.AnnotationParser.parseClass(AnnotationParser.java:663)
		at org.eclipse.jetty.ee10.annotations.AnnotationParser.parseDir(AnnotationParser.java:609)
```

This happens when annotations are scanned asynchronously here:
https://github.com/jetty/jetty.project/blob/9a23f063144c0f70dc1a3d5021759419125cd149/jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationConfiguration.java#L530C25-L543

Because then, `AnnotationConfiguration.DiscoveredServletContainerInitializerHolder.addStartupClasses` is called concurrently from multiple threads, and just adds the provided classes into a `HashSet` without any synchronization.

But HashSet is not thread-safe, so either some thread-safe alternative should be used, or access to it must be synchronized externally.

The workaround is to disable multi-threading for annotation scanning via the "org.eclipse.jetty.annotations.multiThreaded" property.

**How to reproduce?**
The failure happens intermittently, and more often on machines with lots of processors. Also, the probability to run into this is probably higher in large code bases.

--- a/jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationConfiguration.java
+++ b/jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationConfiguration.java
@@ -333,7 +333,10 @@ public class AnnotationConfiguration extends AbstractConfiguration
         @Override
         public void addStartupClasses(String... names)
         {
-            _discoveredClassNames.addAll(Arrays.asList(names));
+            synchronized (_discoveredClassNames)
+            {
+                _discoveredClassNames.addAll(Arrays.asList(names));
+            }
         }
 
         /**
@@ -386,12 +389,15 @@ public class AnnotationConfiguration extends AbstractConfiguration
                         addInheritedTypes(finalClassnames, classMap, classMap.get(c.getName()));
                 }
 
-                for (String classname:_discoveredClassNames)
+                synchronized (_discoveredClassNames)
                 {
-                    //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
-                    finalClassnames.add(classname);
-                    //walk its hierarchy and find all types that extend or implement the class
-                    addInheritedTypes(finalClassnames, classMap, classMap.get(classname));
+                    for (String classname : _discoveredClassNames)
+                    {
+                        //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
+                        finalClassnames.add(classname);
+                        //walk its hierarchy and find all types that extend or implement the class
+                        addInheritedTypes(finalClassnames, classMap, classMap.get(classname));
+                    }
                 }
             }
             

--- a/jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationConfiguration.java
+++ b/jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationConfiguration.java
@@ -313,9 +313,9 @@ public class AnnotationConfiguration extends AbstractConfiguration
     
     public static class DiscoveredServletContainerInitializerHolder extends ServletContainerInitializerHolder
     {
-        private final Set<Class<?>> _handlesTypes = new HashSet<>();
-        private final Set<String> _discoveredClassNames = new HashSet<>();
-        
+        private final Set<Class<?>> _handlesTypes = ConcurrentHashMap.newKeySet();
+        private final Set<String> _discoveredClassNames = ConcurrentHashMap.newKeySet();
+
         public DiscoveredServletContainerInitializerHolder(Source source, ServletContainerInitializer sci, Class<?>... startupClasses)
         {
             super(source, sci);
@@ -333,10 +333,7 @@ public class AnnotationConfiguration extends AbstractConfiguration
         @Override
         public void addStartupClasses(String... names)
         {
-            synchronized (_discoveredClassNames)
-            {
-                _discoveredClassNames.addAll(Arrays.asList(names));
-            }
+            _discoveredClassNames.addAll(Arrays.asList(names));
         }
 
         /**
@@ -389,15 +386,12 @@ public class AnnotationConfiguration extends AbstractConfiguration
                         addInheritedTypes(finalClassnames, classMap, classMap.get(c.getName()));
                 }
 
-                synchronized (_discoveredClassNames)
+                for (String classname : _discoveredClassNames)
                 {
-                    for (String classname : _discoveredClassNames)
-                    {
-                        //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
-                        finalClassnames.add(classname);
-                        //walk its hierarchy and find all types that extend or implement the class
-                        addInheritedTypes(finalClassnames, classMap, classMap.get(classname));
-                    }
+                    //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
+                    finalClassnames.add(classname);
+                    //walk its hierarchy and find all types that extend or implement the class
+                    addInheritedTypes(finalClassnames, classMap, classMap.get(classname));
                 }
             }
             

--- a/jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationConfiguration.java
+++ b/jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationConfiguration.java
@@ -309,8 +309,8 @@ public class AnnotationConfiguration extends AbstractConfiguration
     
     public static class DiscoveredServletContainerInitializerHolder extends ServletContainerInitializerHolder
     {
-        private Set<Class<?>> _handlesTypes = new HashSet<>();
-        private Set<String> _discoveredClassNames = new HashSet<>();
+        private final Set<Class<?>> _handlesTypes = new HashSet<>();
+        private final Set<String> _discoveredClassNames = new HashSet<>();
         
         public DiscoveredServletContainerInitializerHolder(Source source, ServletContainerInitializer sci, Class<?>... startupClasses)
         {
@@ -329,7 +329,10 @@ public class AnnotationConfiguration extends AbstractConfiguration
         @Override
         public void addStartupClasses(String... names)
         {
-            _discoveredClassNames.addAll(Arrays.asList(names));
+            synchronized (_discoveredClassNames)
+            {
+                _discoveredClassNames.addAll(Arrays.asList(names));
+            }
         }
 
         /**
@@ -382,12 +385,15 @@ public class AnnotationConfiguration extends AbstractConfiguration
                         addInheritedTypes(finalClassnames, classMap, (Set<String>)classMap.get(c.getName()));
                 }
 
-                for (String classname:_discoveredClassNames)
+                synchronized (_discoveredClassNames)
                 {
-                    //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
-                    finalClassnames.add(classname);
-                    //walk its hierarchy and find all types that extend or implement the class
-                    addInheritedTypes(finalClassnames, classMap, (Set<String>)classMap.get(classname));
+                    for (String classname : _discoveredClassNames)
+                    {
+                        //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
+                        finalClassnames.add(classname);
+                        //walk its hierarchy and find all types that extend or implement the class
+                        addInheritedTypes(finalClassnames, classMap, (Set<String>)classMap.get(classname));
+                    }
                 }
             }
             

--- a/jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationConfiguration.java
+++ b/jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationConfiguration.java
@@ -306,11 +306,11 @@ public class AnnotationConfiguration extends AbstractConfiguration
             return Integer.compare(i1, i2);
         }
     }
-    
+
     public static class DiscoveredServletContainerInitializerHolder extends ServletContainerInitializerHolder
     {
-        private final Set<Class<?>> _handlesTypes = new HashSet<>();
-        private final Set<String> _discoveredClassNames = new HashSet<>();
+        private final Set<Class<?>> _handlesTypes = ConcurrentHashMap.newKeySet();
+        private final Set<String> _discoveredClassNames = ConcurrentHashMap.newKeySet();
         
         public DiscoveredServletContainerInitializerHolder(Source source, ServletContainerInitializer sci, Class<?>... startupClasses)
         {
@@ -329,10 +329,7 @@ public class AnnotationConfiguration extends AbstractConfiguration
         @Override
         public void addStartupClasses(String... names)
         {
-            synchronized (_discoveredClassNames)
-            {
-                _discoveredClassNames.addAll(Arrays.asList(names));
-            }
+            _discoveredClassNames.addAll(Arrays.asList(names));
         }
 
         /**
@@ -385,18 +382,15 @@ public class AnnotationConfiguration extends AbstractConfiguration
                         addInheritedTypes(finalClassnames, classMap, (Set<String>)classMap.get(c.getName()));
                 }
 
-                synchronized (_discoveredClassNames)
+                for (String classname : _discoveredClassNames)
                 {
-                    for (String classname : _discoveredClassNames)
-                    {
-                        //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
-                        finalClassnames.add(classname);
-                        //walk its hierarchy and find all types that extend or implement the class
-                        addInheritedTypes(finalClassnames, classMap, (Set<String>)classMap.get(classname));
-                    }
+                    //add each of the classes that were discovered to have an annotation listed in @HandlesTypes
+                    finalClassnames.add(classname);
+                    //walk its hierarchy and find all types that extend or implement the class
+                    addInheritedTypes(finalClassnames, classMap, (Set<String>)classMap.get(classname));
                 }
             }
-            
+
             //finally, add the complete set of startup classnames
             super.addStartupClasses(finalClassnames.toArray(new String[0]));
         }

